### PR TITLE
check if tde enabled but extension does not loaded

### PR DESCRIPTION
**Pull request by Sasasu**
## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/utils/fmgr/dfmgr.c [hunk 1]
```diff
@@ -297,6 +298,13 @@ internal_load_library(const char *libname)
 		file_tail = file_scanner;
 	}
 
+	if (file_exists("data_encryption.key") && file_read_buffer_modify_hook == NULL)
+	{
+		ereport(ERROR,
+				(errmsg("this cluster has encryption enabled, but necessary extensions does not loaded"),
+				 errhint("add gp_data_encryption to shared_preload_libraries, in file postgresql.conf")));
```
Comment (beeender) [inline src/backend/utils/fmgr/dfmgr.c]: `errhint("add ‘gp_data_encryption’ to ‘shared_preload_libraries’, in file ‘postgresql.conf’")));`

Add quote `''` or `""` to the terms.

File: src/backend/utils/fmgr/dfmgr.c [hunk 2]
```diff
@@ -297,6 +298,13 @@ internal_load_library(const char *libname)
 		file_tail = file_scanner;
 	}
 
+	if (file_exists("data_encryption.key") && file_read_buffer_modify_hook == NULL)
+	{
+		ereport(ERROR,
+				(errmsg("this cluster has encryption enabled, but necessary extensions does not loaded"),
```
Comment (beeender) [inline src/backend/utils/fmgr/dfmgr.c]: `does not loaded` -> `is not loaded`

File: src/backend/utils/fmgr/dfmgr.c [hunk 3]
```diff
@@ -297,6 +298,13 @@ internal_load_library(const char *libname)
 		file_tail = file_scanner;
 	}
 
+	if (file_exists("data_encryption.key") && file_read_buffer_modify_hook == NULL)
```
Comment (beeender) [inline src/backend/utils/fmgr/dfmgr.c]: Isn't there a better place for this check? Can this be done in the postmaster start, right before loading data files. Wasn't `internal_load_library()` be called when loading any so files? This doesn't look like the right place for the hook.

And, please add a test in TDE side to ensure the behaviour.

File: src/backend/utils/init/miscinit.c [hunk 4]
```diff
@@ -1693,6 +1694,18 @@ process_shared_preload_libraries(void)
...
 	process_shared_preload_libraries_in_progress = false;
+
+	{
+		struct stat st;
+
+		if (file_read_buffer_modify_hook == NULL)
+		{
+			int file_exists = (stat("data_encryption.key", &st) ? (errno == ENOENT) : !S_ISDIR(st.st_mode));
+			ereportif(file_exists, ERROR,
+					(errmsg("this cluster has encryption enabled, but necessary extensions is not loaded"),
```
Comment (beeender) [inline src/backend/utils/init/miscinit.c]: `necessary extensions is` -> `necessary extension is` or `necessary extensions are`

**Review by beeender — APPROVED** [review]
`postinit.c` will call `process_shared_preload_libraries()` as well. If the

File: src/backend/utils/init/miscinit.c [hunk 5]
```diff
@@ -1693,6 +1694,19 @@ process_shared_preload_libraries(void)
 				   false);
 
 	process_shared_preload_libraries_in_progress = false;
+
+	{
+		struct stat st;
+
+		if (file_read_buffer_modify_hook == NULL)
+		{
+			int file_exists = (stat("data_encryption.key", &st) ? (errno == ENOENT) : !S_ISDIR(st.st_mode));
```
Comment (higuoxing) [inline src/backend/utils/init/miscinit.c]: Shall we use an absolute path? e.g.,

```
snprintf(key_path, sizeof(key_path), "%s/data_encryption.key", DataDir);
```

BTW, the logic here doesn't seem quite right.

if `stat()` returns `-1` then `file_exists` will be assigned with `errno == ENOENT`. When `errno` is not `ENOENT` (which could be `EACCES`, `EBADF`, ...), the value of `file_exists` is `false`, and then the error will not be reported.

Comment (Sasasu) [reply src/backend/utils/init/miscinit.c]: > Shall we use an absolute path? e.g.,

No. code under `src/backend/storage` never use absolute path. Only code under `src/bin` is using absolute path.
`src/bin` does not do chdir, but inside PostgreSQL, the current dir always be `DataDir`.

> the logic here doesn't seem quite right.

Changed this part from false negative to false positive.

File: src/backend/utils/init/miscinit.c [hunk 6]
```diff
@@ -1693,6 +1694,22 @@ process_shared_preload_libraries(void)
 				   false);
 
 	process_shared_preload_libraries_in_progress = false;
+
+	{
+		struct stat st;
+
+		if (file_read_buffer_modify_hook == NULL)
+		{
+			int file_exists = (stat("data_encryption.key", &st) ? (errno != ENOENT) : true);
```
Comment (higuoxing) [inline src/backend/utils/init/miscinit.c]: What about the following logic? If we can stat the file or the errno not equals to ENOENT, we assume the file exists.
```c
int file_exists = state("data_encryption.key", &st) == 0 || errno != ENOENT;
```

File: src/backend/utils/init/miscinit.c [hunk 7]
```diff
@@ -1693,6 +1694,22 @@ process_shared_preload_libraries(void)
...
+			errno = 0; // reset errno because we expected errno in stat
```
Comment (higuoxing) [inline src/backend/utils/init/miscinit.c:1704]: It looks that we don't need reset errno since it will be set when error occurs. Right?

**Review by higuoxing — APPROVED** [review]
Looks good with minor comments.

File: src/backend/utils/init/miscinit.c (hunk 7 above)
Comment (Sasasu) [reply src/backend/utils/init/miscinit.c:1704]: PostgreSQL does not reset errno before IO syscall.

```
stat('') -> -1, errno=ENOENT
fopen() -> 0, but errno still = ENOENT
```

Comment (higuoxing) [reply src/backend/utils/init/miscinit.c:1704]: If `fopen()` returns a valid pointer we won't check the `errno`. Only when `fopen()` returns `NULL` we will check `errno` and the `errno` should be set by `fopen()`. Am I right?

**Comment by Sasasu** [comment]
move to github.com/Sasasu/gpdb/pull/12

**Pull request closed without merging**